### PR TITLE
Allow multiple name matchers.

### PR DESCRIPTION
**Pull request by ltratt**
This allows you to specify multiple pattern regex's if you need them. It's probably not going to be used by many people, but it's not hard to imagine it being useful sometimes, and it nicely generalises the API.

This does imply an API change: you can no longer unset a name matcher. Honestly, I can't imagine anyone ever wanted to do that. However, that does mean that this feature will imply a bump to 0.4.

Tested with lang_tester and yksom to make sure that it doesn't break any "normal" usages of fm.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -358,66 +359,72 @@ impl<'a> FMatcher<'a> {
...
+                            if let Some(textm) = text_re.find(text) {
+                                if self.options.distinct_name_matching {
+                                    for (x, y) in names.iter().chain(new_names.iter()) {
+                                        if x != &ptnm.as_str() && y == &textm.as_str() {
+                                            return false;
+                                        }
                                     }
                                 }
-                                Entry::Vacant(_) => match new_names.entry(ptnm.as_str()) {
+                                if textm.start() == textm.end() {
```
Comment (vext01) [inline src/lib.rs:393]: You could shorten this to:
```
assert_eq!(textm.start(),  textm.end(), "Text pattern matched the empty string.");
```

Comment (ltratt) [reply src/lib.rs:393]: I think this should stay a `panic` because it's the result of an error on the user's part. To me an `assert` should reflect an error on my/our part when developing `fm`. There is an element of angels-on-a-pin in this, though, I admit.

**Comment by vext01** [comment]
I assume you are implementing this for TIR tests in `yk`?

Supposing that it doesn't work out for us there, would you still be happy for this feature to be present in fm?

**Comment by ltratt** [comment]
> Supposing that it doesn't work out for us there, would you still be happy for this feature to be present in fm?

Yes, I think `fm` should have been written with multiple name matchers in all along. At first I wasn't sure if it was safe to do so, but I've now realised it is (assuming my thoughts aren't buggy!).

File: src/lib.rs (hunk 1 above)
Comment (vext01) [reply src/lib.rs:393]: So you prefer the semantic meaning of "panic" to "assert". I ask because in Rust an `assert!` is just a dressed up `panic!`.

Comment (ltratt) [reply src/lib.rs:393]: Yes and yes.

Comment (vext01) [reply src/lib.rs:393]: OK

File: src/lib.rs [hunk 2]
```diff
@@ -41,7 +41,7 @@
 /// let text_re = Regex::new(r".+?\b").unwrap();
 /// let matcher = FMBuilder::new("$1 $1")
 ///                         .unwrap()
-///                         .name_matcher(Some((ptn_re, text_re)))
+///                         .name_matcher((ptn_re, text_re))
```
Comment (vext01) [inline src/lib.rs]: Would it be more user friendly for this function to accept two values, rather than a pair?

Comment (ltratt) [reply src/lib.rs]: Yes, probably. It's a mildly more annoying API transition, but if you think it's worth it, I'm happy to make that change.

File: src/lib.rs [hunk 3]
```diff
@@ -358,66 +359,72 @@ impl<'a> FMatcher<'a> {
...
+                        break;
                     }
                 }
-                None => {
-                    if eww {
-                        text.starts_with(&ptn[..ptn.len() - WILDCARD.len()])
-                    } else {
-                        ptn == text
-                    }
+                if (eww && text.starts_with(&ptn[..ptn.len() - WILDCARD.len()])) || ptn == text {
```
Comment (vext01) [inline src/lib.rs:423]: I know this isn't a new variable name, but `eww` is indeed "eww". Should it be renamed (in another PR) as I can't guess what it could mean :P

Same for `sww`.

Comment (ltratt) [reply src/lib.rs:423]: We can argue over the names if you want (I think they're OK if you read the code), but I think we should declare that out-of-scope for this PR.

Comment (vext01) [reply src/lib.rs:423]: Agreed that it's not for this PR, and probably pro-level bikeshedding too.

But honestly `eww` could have been called `dolphin_jimbler` and convey about the same meaning to the casual reader :P

Not a hill I will die on anyway.

Comment (ltratt) [reply src/lib.rs:423]: dolphin_jimbler? :p

File: src/lib.rs (hunk 2 above)
Comment (vext01) [reply src/lib.rs]: Cool. The API already changed here, so you have nothing to lose.

File: src/lib.rs (hunk 3 above)
Comment (vext01) [reply src/lib.rs:423]: Exactly!

File: src/lib.rs (hunk 2 above)
Comment (ltratt) [reply src/lib.rs]: Addressed in https://github.com/softdevteam/fm/pull/24/commits/3f30317a263ba9f1498dde80649b57172c0b6c82.

**Comment by vext01** [comment]
I think we are good to go. Please squash.

**Comment by ltratt** [comment]
I think the commits are best left separate if you're OK with that.
